Feature/travel staff

justliliandev opened the pull request:
# Description

Added Staff of Travelling
Added TeleportationTargetSystem using WorldSavedData (can be moved to capability, if wanted, but WorldSavedData is imo the better approach for this)
Marked as a Draft, because I wanted opinions on:
API-Annotation
TravelRegistry allowing Mods to add their own TravelTargets without a dependency on machines
Usage of IMC vs API anotated methods/class vs API package
Every Travel Target defines how it renders

# Todo

- [ ] GUI of Travel Anchor
- [ ] Staff of Travelling Targetting Travel Targets
- [ ] Maybe Rendering, but I would like to put that on Crazy :)

# Checklist:

- [ ] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

justliliandev committed:
renamed package to match the style of using singular nouns

Also removed unused import

justliliandev committed:
Merge branch 'dev/1.18.x' into feature/travel_staff

justliliandev committed:
Staff of Travelling and basic of a TeleportationTargetSystem (Travel Anchor and Telepad Multiblock)

justliliandev committed:
added accidently removed override annotations

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/client/ClientEventHandler.java:
```diff
@@ -1,20 +1,78 @@
+    }
+
+
+    private static void addParticle(Vec3 pos) {
```
I would rename addParticles and getColor to something like 'addTravelParticles' etc
Depending on much ends up being in this class I would even consider moving the travel stuff into a sperate class compleletly. The way it is it could be pretty hard to find where this functionality is implemented

justliliandev replied:
I've put it there, because I would need another TeleportHandler class just for the client stuff, because of classloading issues. Can extract that to another class called TeleportClientHandler but the code is really seperate anyway, so I don't see a big benefit for extracting it out

CrazyPants replied:
The only value I see is that all the teleport functionality would then be in a single package. Someone looking at the teleport code for the first time might have a hard time finding this functionality.
Just a suggestion though, I am happy for it to stay where it is.

justliliandev replied:
moved it

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/client/renderer/teleportation/RenderTeleportTargets.java:
```diff
@@ -0,0 +1,34 @@
+            Vec3 projectedView = mainCamera.getPosition();
+            poseStack.translate(-projectedView.x, -projectedView.y, -projectedView.z);
+
+            //TODO: Scale, View Bobbing, (FOV/Creative Flight offset stuff)
```
Sure, this stuff was pretty complex to get looking the way I wanted from memory. Was the scaling based on screen pos etc not done in the ported mod?

justliliandev replied:
it wasn't as they use a BlockEntityRenderer for some things, which wouldn't work for us, because we have the big teleporter with infinite range. Also I wanted to get the logic done first, so the Draft would be cleaner

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/handler/travel/TeleportHandler.java:
```diff
@@ -0,0 +1,150 @@
+            return true;
+        if (stack.getItem() instanceof IDarkSteelItem darkSteelItem) {
+            //TODO: Check for upgrade;
+        }
```
Are we going to allow this to be extensible by other mods? I am not fussed if we don't bother with this for now.
We could add an ITraveltem interface (the pair of ITravelTarget). This would need a method like 'isAvailable' to handle whether an upgrade is available or not.
This method also doesn't consider if you have enough energy in the item to teleport but I am guessing that is the correct

justliliandev replied:
The ITravelTarget interface can be used as an optional dependency, because it doesn't need to be implemented on minecraft exposed registry objects like item and blocks. Forcing an interface on Items or Blocks (what ctm is doing rn), would make our mod to be a required dependency. This method is only called for particle rendering and other clients can add their own particles or use our addParticleMethod, if we make it public

CrazyPants replied:
Sounds good

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/handler/travel/TeleportHandler.java:
```diff
@@ -0,0 +1,150 @@
+                    player.teleportTo(eventPos.get().x(), eventPos.get().y(), eventPos.get().z());
+                    player.fallDistance = 0;
+                }
+                player.playNotifySound(SoundEvents.ENDERMAN_TELEPORT, SoundSource.PLAYERS, 1F, 1F);
```
Using this method is the sound played to other people close to the player? You are also playing the sound even if the teleport doesn't happen.

justliliandev replied:
it isn't played to other players, should other players nearby hear it?

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/handler/travel/TeleportHandler.java:
```diff
@@ -0,0 +1,150 @@
+        for (double i = BaseConfig.COMMON.ITEMS.TRAVELLING_BLINK_RANGE.get(); i >= 2; i -= 0.5) {
+            Vec3 v3d = targetVec.add(lookVec.scale(i));
+            target = new BlockPos(Math.round(v3d.x), Math.round(v3d.y), Math.round(v3d.z));
+            Optional<Double> ground = isValidGround(level, target.below());
```
You add the player eye height above but then just use target.below() for the ground check. This will work in most cases but it is probably more correct to subtract eye height again

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/handler/travel/TeleportHandler.java:
```diff
@@ -0,0 +1,150 @@
+     * @param target
+     * @return Optional.empty if it can't teleport and the height where to place the player. This is so you can tp ontop of carpets up to a whole block
+     */
+    private static Optional<Double> isValidGround(BlockGetter level, BlockPos target) {
```
I would rename this to isTelportPositionClear or something. The name sounds like it is testing if the target pos is actual 'ground', not just checking for empty space. On initial reading I though you couldn't teleport to a position in mid air.

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/TravelStaffItem.java:
```diff
@@ -0,0 +1,126 @@
+    @Override
+    public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand usedHand) {
+        ItemStack stack = player.getItemInHand(usedHand);
+        if (getActivationStatus(stack).isAir()) {
```
Both 'use' and 'useOn' should be able to do both teleport types. Most of the time you wish to travel to a block you will right click the air which will come through this method.

CrazyPants commented on enderio-base/src/main/java/com/enderio/base/common/item/tool/TravelStaffItem.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Override
+    public InteractionResult useOn(UseOnContext context) {
```
So you need to make sure the block you are right clicking on hasn't 'consumed' the right click before doing the teleport? Can't remember how this works.

justliliandev replied:
the method is only called if the block hasn't consumed the right click

CrazyPants commented on enderio-machines/src/main/java/com/enderio/machines/common/blockentity/TravelAnchorBlockEntity.java:
```diff
@@ -0,0 +1,85 @@
+
+    public TravelAnchorBlockEntity(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+        super(pType, pWorldPosition, pBlockState);
+        add2WayDataSlot(new StringDataSlot(this::getName, this::setName, SyncMode.GUI));
```
We probably shouldn't be using a number in the middle of a method name like this

justliliandev replied:
do you have a better name for it?

tyler489 replied:
addTwoWayDataSlot

CrazyPants replied:
Yep

justliliandev committed:
Minor changes

Moved ClientEventHandler to seperate Handlers with better names
Renamed isValidGround to a better name
Check for buildheight, so you don't teleport below the world
TravelStaff can teleport to valid TravelTargets
fixed sneak caching
you are now teleported to the center of the block

Rover656 commented:
Closing in favor of #103
